### ISSUE TITLE
Buildings.Templates.AirHandlersFans.Validation.VAVMultiZone fails during NFTyping.typeClassSections

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13115]
@perost please check [Buildings.Templates.AirHandlersFans.Validation.VAVMultiZone](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.AirHandlersFans.Validation.VAVMultiZone.err). The model fails with

```
[Compiler/NFFrontEnd/NFTyping.mo:3019:9-3019:126:writable] Error: 
Internal error NFTyping.typeComponentSections got uninstantiated component capFunFF
```

**Comment by perost** [comment]
MWE:
```Modelica
record PerformanceCurve
  parameter Real capFunFF[:] = {0};
end PerformanceCurve;

record DXCoil
  parameter Integer nSta;
  PerformanceCurve perCur[nSta];
end DXCoil;

record Carrier_Centurion_50PG06 = DXCoil(nSta = 1);

record DataVAVMultiZone
  parameter Integer typCoiCoo;
  parameter Coil coiCoo(typ = typCoiCoo);
end DataVAVMultiZone;

record Coil
  parameter Integer typ;
  parameter Carrier_Centurion_50PG06 datCoi;
end Coil;

model VAVMultiZone
  parameter DataVAVMultiZone dat = VAV;
  parameter Coil coiDat = dat.coiCoo;
  parameter DataVAVMultiZone cfg(typCoiCoo = coiDat.typ);
  parameter DataVAVMultiZone VAV(typCoiCoo = cfg.typCoiCoo);
end VAVMultiZone;
```
I think this model is actually illegal due to circular dependencies between the parameters, but I'm not sure if it's representative of the full model. Regardless it shouldn't be giving internal errors like that.

**Comment by casella** [comment]
@mwetter, FYI

**Comment by mwetter** [comment]
@AntoineGautier: Can you please check if there is indeed a circular dependency in the template model.

**Comment by AntoineGautier** [comment]
## About the circular dependencies

I don't think that the MWE above is representative of the parameter bindings used in `Buildings.Templates.AirHandlersFans.Validation.VAVMultiZone`.

The logic behind these bindings is explained in the paper https://ecp.ep.liu.se/index.php/modelica/article/view/929/837 Section 6.2 and illustrated with a MWE in Listing 2 of this paper.

What stands out in the MWE above is the following declaration:
```mo
  parameter DataVAVMultiZone cfg(typCoiCoo = coiDat.typ);
```
In MBL templates, we rather use two distinct record classes to handle parameter propagation:

- a so-called "configuration" record which stores the structural parameters of all components (such as `typ*`),
- the actual "data" record which stores all the numerical parameters that are used for component sizing and control settings (such as `capFunFF` and `perCur`).

The configuration parameters are propagated ***up*** from the template components to the configuration record of the template, and ultimately to the top-level data record via a composite component binding `dat(cfg=template.cfg)` (where they are used to enable/disable parameters).
The design and operating parameters are propagated ***down*** from the top-level data record to the template components.

Therefore, there shouldn't be any occurrence where a parameter from the configuration record (`cfg.typCoiCoo`) is assigned a value from the data record (`coiDat.typ`).

## About the NFTyping uninstantiated component issue

My guess is that this error is perhaps due to the declaration of `capFunFF[:]` with unknown size. 

I've attached a package that reproduces this error with the same logic for parameter bindings as the one used in MBL templates. (This package also includes the MWE Listing2 from the paper, which is compiled successfully with OMC.)

[TestOpenModelica.tar.gz](https://github.com/user-attachments/files/17729532/TestOpenModelica.tar.gz)

Simulating `NFTypingError` with OMC yields:
```
Internal error NFTyping.typeComponentSections got uninstantiated component mat
```
Now, if we explicitly specify the size of `mat` in `RecordTemplate` as:
```mo
parameter Real mat[cfg.n, cfg.n];
```
the compilation succeeds.

Note that from a template development perspective, it is usually not possible to assign the size of such parameters in the record definition. So, I think that OMC should support the declaration of array parameters with unknown size, with binding equations provided anywhere in the instance tree.

**Comment by perost** [comment]
>So, I think that OMC should support the declaration of array parameters with unknown size, with binding equations provided anywhere in the instance tree.

We already support this, since it's something pretty much every library uses extensively. The issue here is rather that the compiler fails to determine what the size should be. Looking at the model we have:
```Modelica
model NFTypingError
  parameter RecordTemplate datSys(final cfg = sys.cfg, mat = {fill(i/datSys.cfg.n, datSys.cfg.n) for i in 1:datSys.cfg.n});
  Template sys(n = 3, final dat = datSys, com(typ = false));
end NFTypingError;
```
`datSys` and `sys` depend on each other here, and while it's not the type of circular dependency that's forbidden in Modelica it still seems to be causing the compiler some issues. If the modifier for `mat` is changed to `mat = {fill(i/sys.cfg.n, sys.cfg.n) for i in 1:sys.cfg.n}` instead it works fine.

I also tried partially flattening the model by hand and got:
```Modelica
model NFTypingError
  parameter RecordConfiguration datSys_cfg = sys_cfg;
  parameter RecordComponent datSys_com(typ = datSys_cfg.typCom);
  parameter Real datSys_mat[:, :] = {fill(i/datSys_cfg.n, datSys_cfg.n) for i in 1:datSys_cfg.n};
  parameter Integer sys_n = 3;
  parameter RecordConfiguration sys_cfg(n = sys_n, typCom = sys_com.typ);
  parameter RecordConfiguration sys_dat_cfg = datSys_cfg;
  parameter RecordComponent sys_dat_com = datSys_com;
  parameter Real sys_dat_mat[:, :] = datSys_mat;
  Component sys_com(dat = sys_dat_com, typ = false);
end NFTypingError;
```
This also works fine, so my best guess is that the compiler ends up trying to evaluate the whole `datSys` record when deducing the size of `datSys.mat` because it depends on `datSys.cfg.n` and giving up because `datSys` is already in the process of being typed.

We already try to avoid evaluating things unnecessarily when typing `:` dimensions to avoid such issues, but it seems we might need to improve it further to be able to handle cases like this.

**Comment by casella** [comment]
@perost on a scale from 0 to 10 how hard will this be?

**Comment by casella** [comment]
@perost can you please put this again on your agenda for the beginning of 2026? Not urgent, but it needs to be fixed eventually. Or, a workaround for the Buildings library proposed.

**Comment by casella** [comment]
@perost please confirm this is on your to-do list. Thanks!

**Comment by casella** [comment]
@perost ping 😃

**Comment by perost** [comment]
#15843 fixes this and #13116, but this model still fails due to expandable connector issues. But that's a different issue and already covered by other tickets, so we can close this one once #15843 is merged.
